### PR TITLE
Option: Use fused_batch_norm instead of batch_norm for layer normalization

### PR DESCRIPTION
**Pull request by thorjohnsen**
tf.nn.fused_batch_norm uses cudnn for GPU acceleration, tf.nn.batch_normalization does not. This PR adds the option of using fused_batch_norm instead of batch_normalization. It is an opt-in option because fused_batch_norm only improves runtimes on GPUs and does not support double precision tensors.

### CODE REVIEW

**Comment by AzizCode92** [comment]
Thank you for approaching this issue. 
Have you tried this change on a specific task? Did it make the layer normalization faster?

**Comment by thorjohnsen** [comment]
@AzizCode92 I used it in the attention layer of the MLPerf Transformer benchmark. It made the benchmark run about 25% faster. I don't know how familiar you are with MLPerf, it is a google initiative, the public repository is in https://github.com/mlperf/training/. The Transformer reference code has its own implementation of layer normalization (lines 247-265 in file translation/tensorflow/transformer/model/transformer.py), I changed that to use tf.contrib.layers.layer_norm, which simplified the code, but did not make it any faster. I added the option to use tf.nn.fused_batch_norm and got a 25% jump in throughput on GPUs when running our mixed precision version of Transformer. I have not tested what kind of impact this has on the reference code, which uses purely fp32 arithmetic.

**Comment by AzizCode92** [comment]
Interesting!!! I'm sorry I have no idea about the task you're working on.
I'm applying your patch to the  tf.contrib.rnn.LayerNormBasicLSTMCell where they're applying the layer_norm function. Before I noticed that applying the current code, the training time was too slow. 
Have a look at [this issue](https://github.com/tensorflow/tensorflow/issues/21929)
I have also to notice that I combine your modification to this interesting [pull request ](https://github.com/tensorflow/tensorflow/pull/15861).
I'm still training my model and I hope with these two patches I can achieve layerNorm faster and more accurate. 
thank you

**Comment by AzizCode92** [comment]
#Update : it is definitely faster

**Comment by AzizCode92** [comment]
I also recommend you to see this [pull request](https://github.com/MycChiu/fast-LayerNorm-TF)

**Comment by thorjohnsen** [comment]
@AzizCode92 Thanks for telling me about the custom layer norm PR, I'll give that a try.

**Comment by AzizCode92** [comment]
Good, I have worked on it. Unfortunately all the test cases failed!!!
It is an old repo ( 2 years ago ) but worth spending time to fix it

**Comment by AzizCode92** [comment]
#Update : the latest PR I mentioned is working fine now.
Current tensorflow version : 1.10
Cuda v.9
and I can confirm it is faster than the normal layer_norm operation we have now

File: tensorflow/contrib/layers/python/layers/layers.py [hunk 1]
```diff
@@ -2205,7 +2206,8 @@ def layer_norm(inputs,
                trainable=True,
                begin_norm_axis=1,
                begin_params_axis=-1,
-               scope=None):
+               scope=None,
+               use_fused_batch_norm=False):
```
Comment (fchollet) [inline tensorflow/contrib/layers/python/layers/layers.py]: You can just call this argument "fused".

File: tensorflow/contrib/layers/python/layers/layers.py [hunk 2]
```diff
@@ -2307,19 +2311,57 @@ def layer_norm(inputs,
...
+      else:
+        norm_shape = tf.shape(inputs)[begin_norm_axis:]
+        bn_shape = [1, -1, 1, tf.reduce_prod(norm_shape)]
+      if inputs.get_shape().is_fully_defined():
+        outputs_shape = inputs.get_shape()
+      else:
+        outputs_shape = tf.shape(inputs)
+      inputs = array_ops.reshape(inputs, bn_shape)
+      if inputs.get_shape().is_fully_defined():
+        # static inputs TensorShape fully defined after reshape.
```
Comment (fchollet) [inline tensorflow/contrib/layers/python/layers/layers.py]: ?

**Review by fchollet — COMMENTED** [review]
Thanks for the PR.

Please add appropriate unit tests.

Also, note that `contrib.layers` is deprecated and will be removed entirely in TensorFlow 2.0, which is coming soon.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @yifeif: It has been 17 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by yifeif** [comment]
@thorjohnsen could you address @fchollet 's comments? Thanks!

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.
